phpunit/html-api/class-gutenberg-html-tag-processor-6-3-test.php

```diff
@@ -0,0 +1,45 @@
+<?php
+/**
+ * Unit tests covering Gutenberg_HTML_Tag_Processor_6_3 functionality.
+ *
+ * @package WordPress
+ * @subpackage HTML
+ */
+
+require_once __DIR__ . '/../../lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php';
+
+/**
+ * @group html-api
+ *
+ * @coversDefaultClass Gutenberg_HTML_Tag_Processor_6_3
+ */
+class Gutenberg_HTML_Tag_Processor_6_3_Test extends WP_UnitTestCase {
+	/**
+	 * @covers has_bookmark
+	 */
+	public function test_has_bookmark_returns_false_if_bookmark_does_not_exist() {
+		$p = new Gutenberg_HTML_Tag_Processor_6_3( '<div>Test</div>' );
+		$this->assertFalse( $p->has_bookmark( 'my-bookmark' ) );
+	}
+
+	/**
+	 * @covers has_bookmark
+	 */
+	public function test_has_bookmark_returns_true_if_bookmark_exists() {
+		$p = new Gutenberg_HTML_Tag_Processor_6_3( '<div>Test</div>' );
+		$p->next_tag();
+		$p->set_bookmark( 'my-bookmark' );
+		$this->assertTrue( $p->has_bookmark( 'my-bookmark' ) );
+	}
+
+	/**
+	 * @covers has_bookmark
+	 */
+	public function test_has_bookmark_returns_false_if_bookmark_has_been_released() {
+		$p = new Gutenberg_HTML_Tag_Processor_6_3( '<div>Test</div>' );
+		$p->next_tag();
+		$p->set_bookmark( 'my-bookmark' );
+		$p->release_bookmark( 'my-bookmark' );
+		$this->assertFalse( $p->has_bookmark( 'my-bookmark' ) );
+	}
```

Comment:
not we can verify already in a test if a bookmark exists or not, because `$p->seek( 'place' )` will return `false` if it's unavailable.

```php
$this->assertFalse( $p->seek( 'nonexistent' ) );
$this->assertTrue( $p->seek( 'existent' ) );
// wipe out bookmark with some HTML modifications
$this->assertFalse( $p->seek( 'existent' ) );
```

Reply:
Are you suggesting to add a separate test to cover that behavior of `seek`? (Happy to do so 😊 )

Reply:
Ah, I missed the `// wipe out bookmark with some HTML modifications` part. We don't have any API to do that yet though, do we? 🤔

Reply:
> Are you suggesting to add a separate test to cover that behavior of seek?

No; that's a valid thing to talk about, but I was _only_ responding to what I now realize could be a misreading? I thought you had said that it wasn't possible to test _before_ this patch if we have a bookmark or not, and I was suggesting that we _do_, it's just a method that also impacts the processor state.

> We don't have any API to do that yet though, do we? 🤔

That's correct. It's intentionally possible to mess this up though, hopefully in quiet dark corners with blatant warnings to look away. So yes, we could actually write some tests of this behavior before it's necessary, for the purpose of writing tests.

A simpler alternative is simply to assert that a seek to a nonexistent bookmark doesn't exist. I was trying to provide more context in that snippet.

Reply:
> No; that's a valid thing to talk about, but I was only responding to what I now realize could be a misreading? I thought you had said that it wasn't possible to test before this patch if we have a bookmark or not, and I was suggesting that we do, it's just a method that also impacts the processor state.

Right, thanks for clarifying! 

FWIW, part of the reason for adding `has_bookmark` was to [give the user a way to find out](https://github.com/WordPress/gutenberg/pull/47559#issuecomment-1410220038) without triggering a PHP notice when `WP_DEBUG` is set (as you'd [pointed out](https://github.com/WordPress/gutenberg/pull/47559#pullrequestreview-1275954546)).

I'll add a test for `seek( 'nonexistent' )` though!

Reply:
> I'll add a test for `seek( 'nonexistent' )` though!

Hmm, this is a bit more involved. Since the PHP notice _is_ being thrown in unit tests, I need the test to "expect" it. 

I just haven't found the right exception type yet 🤔 

```php
$this->expectException( 'PHPUnit_Framework_Error_Notice' ); // Nope, not that one.
```